File: app/src/main/java/com/arcao/geocaching4locus/authentication/task/GeocachingApiLoginTask.java
package com.arcao.geocaching4locus.authentication.task;

import android.content.Context;

import com.arcao.geocaching.api.GeocachingApi;
import com.arcao.geocaching.api.data.UserProfile;
import com.arcao.geocaching.api.exception.GeocachingApiException;
import com.arcao.geocaching.api.exception.InvalidCredentialsException;
import com.arcao.geocaching.api.exception.InvalidResponseException;
import com.arcao.geocaching4locus.App;
import com.arcao.geocaching4locus.authentication.util.Account;
import com.arcao.geocaching4locus.authentication.util.AccountManager;
import com.arcao.geocaching4locus.authentication.util.DeviceInfoFactory;

public class GeocachingApiLoginTask {
    private final Context context;
    private final GeocachingApi api;

    private GeocachingApiLoginTask(Context context, GeocachingApi api) {
        this.context = context.getApplicationContext();
        this.api = api;
    }

    public static GeocachingApiLoginTask create(Context context, GeocachingApi api) {
        return new GeocachingApiLoginTask(context, api);
    }

    public void perform() throws GeocachingApiException {
        AccountManager manager = App.get(context).getAccountManager();

        Account account = manager.getAccount();
        if (account == null)
            throw new InvalidCredentialsException("Account not found.");

        String token = manager.getOAuthToken();
        if (token == null) {
            manager.removeAccount();
            throw new InvalidCredentialsException("Account not found.");
        }

        api.openSession(token);

        if (manager.isAccountUpdateRequired()) {
            UserProfile userProfile = api.getYourUserProfile(false, false, false, false, false, false, DeviceInfoFactory.create(context));

            if (userProfile == null)
                throw new InvalidResponseException("User profile is null");

            account = manager.createAccount(userProfile.user());
            manager.updateAccount(account);
        }
    }
}

File: app/src/main/java/com/arcao/geocaching4locus/authentication/task/GeocachingApiLoginTask.kt
package com.arcao.geocaching4locus.authentication.task

import android.content.Context
import com.arcao.geocaching.api.GeocachingApi
import com.arcao.geocaching.api.exception.GeocachingApiException
import com.arcao.geocaching.api.exception.InvalidCredentialsException
import com.arcao.geocaching.api.exception.InvalidResponseException
import com.arcao.geocaching4locus.App
import com.arcao.geocaching4locus.authentication.util.Account
import com.arcao.geocaching4locus.authentication.util.DeviceInfoFactory

@Deprecated("Use GeocachingApiLoginUseCase")
class GeocachingApiLoginTask private constructor(context: Context, private val api: GeocachingApi) {
    private val context: Context = context.applicationContext

    @Throws(GeocachingApiException::class)
    fun perform() {
        val accountManager = App[context].accountManager

        var account: Account? = accountManager.account ?: throw InvalidCredentialsException("Account not found.")

        val token = accountManager.oAuthToken
        if (token == null) {
            accountManager.removeAccount()
            throw InvalidCredentialsException("Account not found.")
        }

        api.openSession(token)

        if (accountManager.isAccountUpdateRequired) {
            val userProfile = api.getYourUserProfile(
                false,
                false,
                false,
                false,
                false,
                false,
                DeviceInfoFactory(App.get(context))()
            ) ?: throw InvalidResponseException("User profile is null")

            account = accountManager.createAccount(userProfile.user())
            accountManager.updateAccount(account)
        }
    }

    companion object {
        @JvmStatic
        fun create(context: Context, api: GeocachingApi): GeocachingApiLoginTask {
            return GeocachingApiLoginTask(context, api)
        }
    }
}
